Adds gitlab output support via --gitlab flag

chgg-kboberg opened the pull request:
Attn: @underyx 

Adds gitlab-formatted output for the Gitlab action.  Should (if I understand _Click behavior correctly) automagically set gitlab output if run in a Gitlab environment.

Also comes with some basic bootstrapping of test infrastructure!

chgg-kboberg committed:
add gitlab output settings

chgg-kboberg committed:
working e2e jsonschema validation

chgg-kboberg committed:
add some comments w/attribution and explanation of hackery

underyx commented on pyproject.toml:
```diff
@@ -18,6 +18,9 @@ gitpython = "^2.1.15"
 attrs = "~=19.3.0"
+pytest = "^6.1.2"
+pathlib = "^1.0.1"
+jsonschema = "^3.2.0"
```
could you install these as dev dependencies instead (`poetry add --dev`)? also, pathlib is already in the standard library of the Python version we use, so you might be able to skip it

underyx commented on src/semgrep_agent/findings.py:
```diff
@@ -98,6 +99,61 @@ def to_dict(self, omit: Set[str]) -> Mapping[str, Any]:
+            "severity": self._to_gitlab_severity(),
+            # KB note:
+            # Semgrep is designed to be a low-FP tool by design.
+            # Does hard-coding confidence make sense here?
```
yep, we did this for SARIF as well

although now I think we should've done `"High" if "audit" not in rule_id.split(".") else "Low"` (the audit rules are to be used with human reviewers checking all results)

chgg-kboberg committed:
fix linter errors

underyx commented on src/semgrep_agent/main.py:
```diff
@@ -142,6 +146,14 @@ def main(
+            "version": "2.0",
+            "vulnerabilities": [f.to_gitlab() for f in new_findings],
+        }
+        click.echo(json.dumps(output))
```
We could even write it directly to a file maybe, if that's what GitLab needs and skip using a pipe. Or do they understand results from stdout as well?

chgg-kboberg replied:
File support is an option.  Gitlab artifact definitions look like the following:

```yaml
goose-sast:
  extends: .sast-analyzer
  image: kboberg/goose:latest
  tags:
    - security
  artifacts:
    reports:
      sast: gl-sast-report.json
  script:
    - /app/goose -target $CI_PROJECT_DIR -config /app/config/goose_rules.json -decisiontree -outputmode gitlab > gl-sast-report.json
````

underyx replied:
cool, let's leave it this way then if other tools tend to use `cmd > gl-sast-report.json` syntax as well 👍

chgg-kboberg committed:
move test-related dependencies to Poetry dev dependencies

underyx commented on pytest/conftest.py:
```diff
@@ -0,0 +1,94 @@
+        stderr=subprocess.STDOUT if stderr else subprocess.PIPE,
+        stdout=subprocess.PIPE
+    )
+    # we have to use Popen,wait,communicate here because findings trigger non-zero exit codes
```
I think `subprocess.run` would have a nicer API that doesn't raise on non-zero exit codes.

underyx replied:
Even better, we could:

- pass a `cwd` parameter there with the directory that has that `stupid.py` with known semgrep issues
- and we can rename the `pytest/` directory back to a more standard `tests/` (it makes me uncomfortable, I worry somehow `import pytest` will think this is the correct module directory 😅 )

chgg-kboberg committed:
use run() instead of Popen()

chgg-kboberg committed:
add pre-commit hook installation docs from core Semgrep

underyx commented on CONTRIBUTING.md:
````diff
@@ -14,6 +14,19 @@ Install dependencies with poetry
+```
+python -m pip install pre-commit
+pre-commit install
+```
````
😻 thanks so much!

underyx commented:
"semgrep with managed policy" only fails cause your fork doesn't have access to our API token by the way, don't worry about that job

chgg-kboberg committed:
move pytest content back under tests, use working directory to move agent inside tests/ to bypass .semgrepignore

chgg-kboberg committed:
move pytest content back under tests, use working directory to move agent inside tests/ to bypass .semgrepignore

chgg-kboberg commented:
Tests now live in the `tests/` directory once more!

![image](https://user-images.githubusercontent.com/56612207/99737138-8922c280-2a7c-11eb-877b-19233fefd7a2.png)

underyx committed:
Reformat code with pre-commit

underyx committed:
Sync .pre-commit-config with r2c/semgrep repo

underyx committed:
Fix type safety issues

underyx committed:
Stop always just generating JSON when in GitLab CI

Using this env var made it impossible to use semgrep-agent in any other
way than for GitLab JSON output without messing with system env vars.

underyx commented:
Awesome job, can't thank you enough! Especially for the testing infrastructure ☺️
